test/test-agent.cpp

```diff
@@ -0,0 +1,69 @@
+#include <gtest/gtest.h>
+
+#include "agent.h"
+
+TEST(AgentTest, Basic)
+{
+    Rep::Agent agent = Rep::Agent().allow("/").disallow("/foo");
+    EXPECT_EQ(agent.directives().size(), 2);
+}
+
+TEST(AgentTest, ChecksAllowed)
+{
+    Rep::Agent agent = Rep::Agent().allow("/path");
+    EXPECT_TRUE(agent.allowed("/path"));
+    EXPECT_TRUE(agent.allowed("/elsewhere"));
+}
+
+TEST(AgentTest, HonorsLongestFirstPriority)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/path").allow("/path/exception");
+    EXPECT_TRUE(agent.allowed("/path/exception"));
+    EXPECT_FALSE(agent.allowed("/path"));
+}
+
+TEST(AgentTest, RobotsTextAllowed)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/robots.txt");
+    EXPECT_TRUE(agent.allowed("/robots.txt"));
+}
+
+TEST(AgentTest, DisallowNone)
+{
+    Rep::Agent agent = Rep::Agent().disallow("");
+    EXPECT_TRUE(agent.allowed("/anything"));
+}
+
+TEST(AgentTest, EscapedRule)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/a%3cd.html");
```

Comment:
The priority is taken after escaping. I don't know which is better, honestly. I think it's rare enough and also kind of a toss-up.